File: String/P539MinimumTimeDifference/MinimumTimeDifference.java
package String.P539MinimumTimeDifference;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

/**
 * @CLassName MinimumTimeDifference
 * @Description TODO
 * @Author cst
 * @Date 2/8/21  11:07 AM
 */
public class MinimumTimeDifference {
    public int findMinDifference(List<String> timePoints) {
        // like find the shortest distances between two elements in a circular array
        boolean[] points = new boolean[1440];
        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;
        for (String time : timePoints) {
            String[] arrs = time.split(":");
            int h =  Integer.parseInt(arrs[0]);
            int m = Integer.parseInt(arrs[1]);
            int t = h * 60 + m;
            // it was visited before, so there must be two exactly same time. we could return 0 directly
            if (points[t]) {
                return 0;
            }
            points[t] = true;
            min = Math.min(min, t);
            max = Math.max(max, t);
        }

        int res = Integer.MAX_VALUE;
        int prev = 0;
        for (int i = min; i<= max; i++ ){
            if (points[i]) {
                if (i == min) {
                    // 比如23：59 和 0:00, 倒过来更接近
                    res = Math.min(res, Math.min(max - min, 1440 - (max - min)));
                } else {
                    res = Math.min(res, i - prev);
                }
                prev = i;
            }
        }
        return res;
    }
}

File: String/P539MinimumTimeDifference/MinimumTimeDifference.py
import unittest
from typing import List
import sys

class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        points, prev = [0] * 1440, 0
        minVal, maxVal = sys.maxsize, -sys.maxsize - 1  
        res = sys.maxsize
        for time in timePoints:
            arr = time.split(":")
            temp = int(arr[0]) * 60 + int(arr[1])
            # we have exactly two same time points
            if points[temp] != 0:
                return 0
            points[temp] = 1
            minVal = min(minVal, temp)
            maxVal = max(maxVal, temp)
        for i in range(minVal, maxVal + 1):
            if points[i] != 0:
                res = min(res, min(maxVal - minVal, 1440 - maxVal + minVal)) if i == minVal else min(res, i - prev)
                prev = i
        return res 
class UnitTest(unittest.TestCase):
    def test(self):
        # timePoints = ["00:00","23:59","00:00"]
        # self.assertEqual(0, Solution().findMinDifference(timePoints))
        
        timePoints = ["23:59","00:00"]
        self.assertEqual(1, Solution().findMinDifference(timePoints))
if __name__ == "__main__":
    unittest.main()   
            